### ISSUE TITLE
Issue with Prepared Statements keep growing

### ISSUE THREAD

**Issue by byronpc** [elixir-ecto/myxql#103]
I'm not sure if this is related to issue #80 but this problem exists on the latest 0.3.3 myxql

But there seems to be 2 issues here related to custom queries 

1. If I send a query that fails a foreign key constraint, the query gets prepared. And if i repeat the same query, it gets prepared again. So the connection keeps alot of copies of the same prepared statements and eventually fills up the maximum prepared statements allowed by mysql

```
Repo.query("INSERT INTO profile_view(viewer_id, viewed_id) VALUES (?, ?)", [1, 1])
{:error,
 %MyXQL.Error{
   connection_id: 11,
   message: "(1452) (ER_NO_REFERENCED_ROW_2) Cannot add or update a child row: a foreign key constraint fails (`db`.`profile_view`, CONSTRAINT `FK_4835241A192BCF11` FOREIGN KEY (`viewed_id`) REFERENCES `user_abstract` (`id`))",
   mysql: %{code: 1452, name: :ER_NO_REFERENCED_ROW_2},
   statement: "INSERT INTO profile_view(viewer_id, viewed_id, created) VALUES (?, ?, now())"
 }}
```

```
select STATEMENT_ID, SQL_TEXT from prepared_statements_instances;
+--------------+------------------------------------------------------------------------------+
| STATEMENT_ID | SQL_TEXT                                                                     |
+--------------+------------------------------------------------------------------------------+
|            3 | INSERT INTO profile_view(viewer_id, viewed_id, created) VALUES (?, ?, now()) |
|            4 | INSERT INTO profile_view(viewer_id, viewed_id, created) VALUES (?, ?, now()) |
|            5 | INSERT INTO profile_view(viewer_id, viewed_id, created) VALUES (?, ?, now()) |
|            6 | INSERT INTO profile_view(viewer_id, viewed_id, created) VALUES (?, ?, now()) |
|            7 | INSERT INTO profile_view(viewer_id, viewed_id, created) VALUES (?, ?, now()) |
+--------------+------------------------------------------------------------------------------+
```

2. Second issue is the if the query succeeds on the other hand, the query doesn't get prepared at all.

```
Repo.query("INSERT INTO profile_view(viewer_id, viewed_id, created) VALUES (?, ?, now())", [3250, 1192152])
{:ok,
 %MyXQL.Result{
   columns: nil,
   connection_id: 12,
   last_insert_id: 43775,
   num_rows: 1,
   num_warnings: 0,
   rows: nil
 }}
```

Checking the prepared_statements_instances, nothing appears

**Comment by josevalim** [comment]
`2` happens because we close the prepared statement after the query. `1` is definitely a bug though.

**Comment by josevalim** [comment]
`2` happens because we close the prepared statement after the query. `1` is definitely a bug though.

**Comment by byronpc** [comment]
Hmmm why prepare it in the first place if you're just gonna close it afterwards? Why not send the query straight as text or maintain the prepared statement? Wouldn't  that be counter performance to do itnlike that? @josevalim

**Comment by wojtekmach** [comment]
Prepared statements use the binary protocol which is much more efficient than the text one, so even if we do multiple hops it tends to be faster especially as size of response grows. You could opt out of that by setting query_type: :text option on MyXQL.query call.

**Comment by byronpc** [comment]
alright understood. but is there a way for us to prepare the statement manually?

**Comment by wojtekmach** [comment]
Yes, with MyXQL.prepare.

**Comment by byronpc** [comment]
cool thanks!

**Comment by byronpc** [comment]
just fyi for others who are also looking into this, i looked into the lib more...we just need to add the `cache_statement: "name"` opt in the `query()` function. so no need to manually execute MyXQL.prepare

**Comment by byronpc** [comment]
@josevalim please reopen this. it seems it still exists. not on manual `query()` but rather via ecto `insert_all`. If the query fails for reasons like foreign key constraints, the prepare statement will keep growing.

e.g.
```
** (MyXQL.Error) (1452) (ER_NO_REFERENCED_ROW_2) Cannot add or update a child row: a foreign key constraint fails (`goodwall`.`profile_view`, CONSTRAIN
T `FK_4835241A192BCF11` FOREIGN KEY (`viewed_id`) REFERENCES `user_abstract` (`id`))
```

ran it multiple times and this is what i got

```
*************************** 3. row ***************************
SQL_TEXT: INSERT INTO `profile_view` (`created`,`viewed_id`,`viewer_id`) VALUES (?,?,?),(?,?,?),(?,?,?)
*************************** 4. row ***************************
SQL_TEXT: INSERT INTO `profile_view` (`created`,`viewed_id`,`viewer_id`) VALUES (?,?,?),(?,?,?),(?,?,?)
*************************** 5. row ***************************
SQL_TEXT: INSERT INTO `profile_view` (`created`,`viewed_id`,`viewer_id`) VALUES (?,?,?),(?,?,?),(?,?,?)
*************************** 6. row ***************************
SQL_TEXT: INSERT INTO `profile_view` (`created`,`viewed_id`,`viewer_id`) VALUES (?,?,?),(?,?,?),(?,?,?)
```

And on the side of that, is there a way to issue an `INSERT IGNORE` here?

**Comment by byronpc** [comment]
sorry will investigate further. not the reason but there's a leak somewhere else

**Comment by byronpc** [comment]
fixed on latest release 0.4.0
